src/EditorFeatures/Core.Wpf/LineSeparators/LineSeparatorAdornmentManager.cs

```diff
@@ -13,50 +13,21 @@
 
 namespace Microsoft.CodeAnalysis.Editor.LineSeparators
 {
-    internal class LineSeparatorAdornmentManager : AdornmentManager<LineSeparatorTag>
+    internal class LineSeparatorAdornmentManager : AbstractAdornmentManager<LineSeparatorTag>
     {
         public LineSeparatorAdornmentManager(IThreadingContext threadingContext, IWpfTextView textView,
             IViewTagAggregatorFactoryService tagAggregatorFactoryService, IAsynchronousOperationListener asyncListener, string adornmentLayerName)
             : base(threadingContext, textView, tagAggregatorFactoryService, asyncListener, adornmentLayerName)
         {
         }
 
-        /// <summary>
-        /// MUST BE CALLED ON UI THREAD!!!!   This method touches WPF.
-        /// 
-        /// This is where we apply visuals to the text. 
-        /// 
-        /// It happens when another region of the view becomes visible or there is a change in tags.
-        /// For us the end result is the same - get tags from tagger and update visuals correspondingly.
-        /// </summary>        
-        protected override void UpdateSpans_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection, bool removeOldTags)
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
         {
-            Contract.ThrowIfNull(changedSpanCollection);
-
             // this method should only run on UI thread as we do WPF here.
             Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
 
             var viewSnapshot = TextView.TextSnapshot;
-            var visualSnapshot = TextView.VisualSnapshot;
-
             var viewLines = TextView.TextViewLines;
-            if (viewLines == null || viewLines.Count == 0)
```

Comment:
Shouldn't we still have these checks?  Or is that already handled elsewhere?

Reply:
Handled at the AbstractAdornmentManager